### ISSUE TITLE
Problem with Accuracies after training. 

### ISSUE THREAD

**Issue by simonefrancia** [keras-team/keras#12308]
Hi, 
I am facing with a strange behaviour with keras after my training for mood classification. 

After I finished my training pipeline, I get an accuracy of 80% on validation for the best epoch, and if I launch the evaluation on test set on the same session, I get similar results like 78% and it's OK for me. 

Now my problem: when I try to save my model and not perform the test set in the same script, but in a different script using load_model, I get completely different and bad result on the same test set and with the same code. 

Here is how I launch my training and how I save my model:

```
script1
model = Model(inputs=[input], outputs=predict)

model.compile(loss='mse', optimizer='adam')

with tf.Session() as session:

    K.set_session(session)
    session.run(tf.global_variables_initializer())
    session.run(tf.tables_initializer())
  
    history = model.fit_generator(training_generator,
                                  steps_per_epoch=len(partition['train'])/batch_size,
                                  callbacks=[csv_logger],#,checkpoint],
                                  epochs=EPOCHS,
                                  verbose=1,
                                  class_weight=None,
                                  max_queue_size=10,
                                  workers=1,
                                  use_multiprocessing=False,
                                  shuffle=True,
                                  initial_epoch=0)
    
    model.save(output_path + '/model_ai_training.h5')  # creates a HDF5 file 'my_model.h5'

```
Then, if script finishes and I try to load the model from a new script and a new session:

```
'script2'
with tf.Session() as session:

    K.set_session(session)
    session.run(tf.global_variables_initializer())
    session.run(tf.tables_initializer())

    model = load_model(output_path + '/model_ai_training.h5' , custom_objects={ "tf" : tf})
    predicts = model.predict(X_test, batch_size=16)

```
Doing this, I get bad results on test set. 
The strange thing is that if I put these two lines at the bottom of script1 these two lines without closing the session:

```
model = load_model(output_path + '/model_ai_training.h5' , custom_objects={ "tf" : tf})
predicts = model.predict(X_test, batch_size=16)
``` 
Results are really better here. 

Thanks

**Comment by loretoparisi** [comment]
@gabrieldemarmiesse is this a possibile issue or maybe a bug on the implementation?

**Comment by gabrieldemarmiesse** [comment]
I Think people should avoid manipulating sessions and things like tf.global_variables_initializer()
and session.run(tf.tables_initializer() when working with keras (because keras uses this in the background and global variables are messy). TF Sessions are really buggy and very error prone in general anyway, so I recommend avoid using them and switch to TF 2 as soon as possible.

**Comment by loretoparisi** [comment]
@gabrieldemarmiesse porting to TF 2.0 it's for sure something that will help. That said, considering these issues, what could it be a possibile workaround in the meanwhile? Thanks.

**Comment by gabrieldemarmiesse** [comment]
I don't think modifying the session variable directly is supported by keras. What you want to achieve should be possible in another (and cleaner) way since keras can make arbitrary direct acyclic graphs. So I don't really see a use case where you would be forced to interact with the session directly.

**Comment by gabrieldemarmiesse** [comment]
Since I cannot see what is in the network it is impossible for me to help.

**Comment by rohit-gupta** [comment]
@gabrieldemarmiesse I am not doing this session stuff, but I am getting the same problem on a different dataset and architecture. I was informed elsewhere that this could be due to an issue with how learning phase interacts with BatchNorm and that did help slightly but didn't solve it. 

In this chart x-axis = epoch number, y-axis = validation set accuracy (same dataset and generator)

![image](https://user-images.githubusercontent.com/825799/59951287-116ef380-9496-11e9-9a88-102c1c18b157.png)
